File: packages/cli/src/server/request-context.ts
import type http from "node:http";
import net from "node:net";
import {
  isTrustedProxy,
  type TrustedProxyConfig,
} from "#app/server/trusted-proxies.ts";

export type RequestContext = {
  readonly clientIp: string;
  readonly host: string | undefined;
  readonly protocol: "http" | "https";
  readonly trustedProxy: boolean;
};

const firstHeaderValue = (
  value: string | string[] | undefined,
): string | undefined => {
  const first = Array.isArray(value) ? value[0] : value;
  const trimmed = first?.trim();
  return trimmed ? trimmed : undefined;
};

const firstCommaValue = (
  value: string | string[] | undefined,
): string | undefined => {
  const first = firstHeaderValue(value)?.split(",")[0]?.trim();
  return first ? first : undefined;
};

const firstValidForwardedIp = (
  value: string | string[] | undefined,
): string | undefined => {
  const values = Array.isArray(value) ? value : value ? [value] : [];

  for (const item of values) {
    for (const candidate of item.split(",")) {
      const ip = candidate.trim();
      if (net.isIP(ip) !== 0) {
        return ip;
      }
    }
  }

  return undefined;
};

const validForwardedIps = (
  value: string | string[] | undefined,
): readonly string[] => {
  const values = Array.isArray(value) ? value : value ? [value] : [];

  return values.flatMap((item) =>
    item
      .split(",")
      .map((candidate) => candidate.trim())
      .filter((candidate) => net.isIP(candidate) !== 0),
  );
};

const clientIpFromForwardedChain = (
  value: string | string[] | undefined,
  trustedProxies: TrustedProxyConfig,
): string | undefined => {
  const ips = validForwardedIps(value);
  if (ips.length === 0) {
    return undefined;
  }

  for (let index = ips.length - 1; index >= 0; index -= 1) {
    const ip = ips[index];
    if (ip && !isTrustedProxy(ip, trustedProxies)) {
      return ip;
    }
  }

  return ips[0];
};

const forwardedProtocol = (
  value: string | string[] | undefined,
  defaultProtocol: "http" | "https",
): "http" | "https" => {
  const protocol = firstCommaValue(value)?.toLowerCase();
  return protocol === "http" || protocol === "https"
    ? protocol
    : defaultProtocol;
};

export const getRequestContext = ({
  defaultProtocol,
  headers,
  remoteAddress,
  trustedProxies,
}: {
  readonly defaultProtocol: "http" | "https";
  readonly headers: http.IncomingHttpHeaders;
  readonly remoteAddress: string | undefined;
  readonly trustedProxies: TrustedProxyConfig;
}): RequestContext => {
  const trustedProxy = isTrustedProxy(remoteAddress, trustedProxies);
  if (!trustedProxy) {
    return {
      clientIp: remoteAddress ?? "",
      host: firstHeaderValue(headers.host),
      protocol: defaultProtocol,
      trustedProxy,
    };
  }

  return {
    clientIp:
      clientIpFromForwardedChain(headers["x-forwarded-for"], trustedProxies) ??
      firstValidForwardedIp(headers["x-real-ip"]) ??
      remoteAddress ??
      "",
    host:
      firstHeaderValue(headers["x-forwarded-host"]) ??
      firstHeaderValue(headers.host),
    protocol: forwardedProtocol(headers["x-forwarded-proto"], defaultProtocol),
    trustedProxy,
  };
};

File: lib/src/services/server/request_context.dart
import 'dart:io';

import 'package:proxer/src/services/server/trusted_proxies.dart';

class RequestContext {
  const RequestContext({
    required this.clientIp,
    required this.host,
    required this.protocol,
    required this.trustedProxy,
  });

  final String clientIp;
  final String? host;
  final String protocol;
  final bool trustedProxy;
}

String? _firstHeader(HttpHeaders headers, String name) {
  final value = headers.value(name)?.split(',').first.trim();
  return value == null || value.isEmpty ? null : value;
}

String? _firstValidIp(Iterable<String> values) {
  for (final value in values) {
    for (final candidate in value.split(',')) {
      final ip = candidate.trim();
      try {
        InternetAddress(ip);
        return ip;
      } on ArgumentError {
        // Continue to the next forwarded value.
      }
    }
  }
  return null;
}

String? _clientFromChain(
  Iterable<String> values,
  TrustedProxyConfig trustedProxies,
) {
  final ips = <String>[];
  for (final value in values) {
    for (final candidate in value.split(',')) {
      final ip = candidate.trim();
      try {
        InternetAddress(ip);
        ips.add(ip);
      } on ArgumentError {
        // Ignore malformed forwarded entries.
      }
    }
  }
  if (ips.isEmpty) {
    return null;
  }
  for (final ip in ips.reversed) {
    if (!isTrustedProxy(ip, trustedProxies)) {
      return ip;
    }
  }
  return ips.first;
}

RequestContext getRequestContext({
  required HttpRequest request,
  required TrustedProxyConfig trustedProxies,
  String defaultProtocol = 'http',
}) {
  final remote = request.connectionInfo?.remoteAddress.address;
  final trusted = isTrustedProxy(remote, trustedProxies);
  if (!trusted) {
    return RequestContext(
      clientIp: remote ?? '',
      host: _firstHeader(request.headers, HttpHeaders.hostHeader),
      protocol: defaultProtocol,
      trustedProxy: false,
    );
  }
  final forwardedProto = _firstHeader(
    request.headers,
    'x-forwarded-proto',
  )?.toLowerCase();
  return RequestContext(
    clientIp:
        _clientFromChain(
          request.headers['x-forwarded-for'] ?? const [],
          trustedProxies,
        ) ??
        _firstValidIp(request.headers['x-real-ip'] ?? const []) ??
        remote ??
        '',
    host:
        _firstHeader(request.headers, 'x-forwarded-host') ??
        _firstHeader(request.headers, HttpHeaders.hostHeader),
    protocol: forwardedProto == 'https' || forwardedProto == 'http'
        ? forwardedProto!
        : defaultProtocol,
    trustedProxy: true,
  );
}
